### PR TITLE
Transaction id

### PR DESCRIPTION
**Pull request by mjburghoffer**
making transaction id better, faster, sortable, more unique, and immune to case-insensitivity

### CODE REVIEW

**Comment by mjburghoffer** [comment]
Perhaps we should also use the machine name/address as some entropy as well so this can scale horizontally a little better?

**Comment by christydennison** [comment]
Sure, sounds good to me if you want to. 200 mill is a lot of transactions, but it would suck if you were that one guy.

**Comment by jontg** [comment]
LGTM!  How do we want to review and merge each of these PRs?  Just LGTM them all as appropriate, and have @mjburghoffer merge everything at once when everything is good to go?

**Comment by christydennison** [comment]
Sounds good to me @jontg !

**Comment by mjburghoffer** [comment]
@christydennison i tweaked this a bit so TransactionIds are only 12 chars long, and are also scoped to machine + process

File: soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+      } catch (Exception ex) {
+        log.warn("Could not compute process name from getRuntimeMXBean");
+        return UUID.randomUUID().toString();
+      }
+    }
+
+    private static String getDefaultJvmName() {
+      try {
+        return TransactionIdProvider.class.getClassLoader().toString();
+      } catch (Exception ex) {
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java]: CRITICAL Either log or rethrow this exception. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)

File: soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    /*
+    0011111111111111111111111111111111111111111111111111111111111111 WHOLE SHEBANG
+    0011111111111111111111111111111111000000000000000000000000000000 TIMESTAMP
+    0000000000000000000000000000000000111111110000000000000000000000 MACHINE
+    0000000000000000000000000000000000000000001111000000000000000000 PROCESS
+    0000000000000000000000000000000000000000000000111111111111111111 COUNTER
+    */
+
+    long timestamp = (transactionId & 0b0011111111111111111111111111111111000000000000000000000000000000L) >>> 30;
+    short machineBits = (short) ((transactionId & 0b0000000000000000000000000000000000111111110000000000000000000000) >>> 22);
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java]: MAJOR Rename "machineBits" which hides the field declared at line 61. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AHiddenFieldCheck)

File: soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+
+  // default process name
+  static class ProcessNameHolder {
+    @Inject(optional = true) @Named("transactionId.process.name")
+    public String value = getDefaultProcessName() + ";" + getDefaultJvmName();
+
+    private static String getDefaultProcessName() {
+      try {
+        return ManagementFactory.getRuntimeMXBean().getName();
+      } catch (Exception ex) {
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java]: CRITICAL Either log or rethrow this exception. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)

File: soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    0011111111111111111111111111111111111111111111111111111111111111 WHOLE SHEBANG
+    0011111111111111111111111111111111000000000000000000000000000000 TIMESTAMP
+    0000000000000000000000000000000000111111110000000000000000000000 MACHINE
+    0000000000000000000000000000000000000000001111000000000000000000 PROCESS
+    0000000000000000000000000000000000000000000000111111111111111111 COUNTER
+    */
+
+    long timestamp = (transactionId & 0b0011111111111111111111111111111111000000000000000000000000000000L) >>> 30;
+    short machineBits = (short) ((transactionId & 0b0000000000000000000000000000000000111111110000000000000000000000) >>> 22);
+    short processBits = (short) ((transactionId & 0b0000000000000000000000000000000000000000001111000000000000000000) >>> 18);
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java]: MAJOR Rename "processBits" which hides the field declared at line 62. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AHiddenFieldCheck)

File: soa-commons/src/main/java/com/relateiq/soa/commons/TransactionId.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    if (comp != 0) {
+      return comp;
+    }
+
+    return this.counterMask - that.counterMask;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) return true;
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionId.java]: MINOR At most one statement is allowed per line, but 2 statements were found on this line. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00122)

File: soa-commons/src/main/java/com/relateiq/soa/commons/TransactionId.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    if (o == null || getClass() != o.getClass()) return false;
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionId.java]: MINOR At most one statement is allowed per line, but 2 statements were found on this line. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00122)

File: soa-commons/src/main/java/com/relateiq/soa/commons/TransactionId.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+    TransactionId that = (TransactionId) o;
+
+    if (timestamp != that.timestamp) return false;
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionId.java]: MINOR At most one statement is allowed per line, but 2 statements were found on this line. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00122)

File: soa-commons/src/main/java/com/relateiq/soa/commons/TransactionId.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    if (machineBits != that.machineBits) return false;
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionId.java]: MINOR At most one statement is allowed per line, but 2 statements were found on this line. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00122)

File: soa-commons/src/main/java/com/relateiq/soa/commons/TransactionId.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    if (processBits != that.processBits) return false;
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionId.java]: MINOR At most one statement is allowed per line, but 2 statements were found on this line. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00122)

**Comment by relateiq-ec2** [comment]
SonarQube analysis reported 9 issues:
* CRITICAL 2 critical
* MAJOR 2 major
* MINOR 5 minor

Watch the comments in this conversation to review them.

**Comment by relateiq-ec2** [comment]
SonarQube analysis reported 7 issues:
* CRITICAL 2 critical
* MINOR 5 minor

Watch the comments in this conversation to review them.

File: soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+      } catch (Exception ex) {
+        log.warn("Could not compute process name from getRuntimeMXBean");
+      }
+      return UUID.randomUUID().toString();
+    }
+
+    private static String getDefaultJvmName() {
+      try {
+        return TransactionIdProvider.class.getClassLoader().toString();
+      } catch (Exception ex) {
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java]: CRITICAL Either log or rethrow this exception. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)

File: soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+
+  // default process name
+  static class ProcessNameHolder {
+    @Inject(optional = true) @Named("transactionId.process.name")
+    public String value = getDefaultProcessName() + ";" + getDefaultJvmName();
+
+    private static String getDefaultProcessName() {
+      try {
+        return ManagementFactory.getRuntimeMXBean().getName();
+      } catch (Exception ex) {
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java]: CRITICAL Either log or rethrow this exception. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)

**Comment by relateiq-ec2** [comment]
SonarQube analysis reported 7 issues:
* CRITICAL 2 critical
* MINOR 5 minor

Watch the comments in this conversation to review them.

File: soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+ * <p>
+ * As such, this should be sufficiently unique and sortable for the time being.
+ */
+public class TransactionIdProvider {
+
+  private static final Logger log = LoggerFactory.getLogger(TransactionIdProvider.class);
+
+  public static final long EPOCH_OFFSET_SECONDS = Duration.between(
+      Instant.EPOCH,
+      LocalDateTime.of(2000, 1, 1, 0, 0).toInstant(ZoneOffset.UTC)
```
Comment (christydennison) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java]: 2000? 2016 seems reasonable :)

File: soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+  public TransactionId decodeTransactionIdFromLong(long transactionId) {
+    /*
+    0011111111111111111111111111111111111111111111111111111111111111 WHOLE SHEBANG
+    0011111111111111111111111111111111000000000000000000000000000000 TIMESTAMP
+    0000000000000000000000000000000000111111000000000000000000000000 MACHINE
+    0000000000000000000000000000000000000000111111000000000000000000 PROCESS
+    0000000000000000000000000000000000000000000000111111111111111111 COUNTER
+    */
+
+    long timestampValue = (transactionId & 0b0011111111111111111111111111111111000000000000000000000000000000L) >>> 30;
```
Comment (christydennison) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java]: maybe have these offsets as constants?

File: soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+ * (2) if there are more than 262,144 transactions per second,
+ * in a process on a machine that hashes to the same hashcode as another, there will be duplicate id's
+ * <p>
+ * As such, this should be sufficiently unique and sortable for the time being.
+ */
+public class TransactionIdProvider {
+
+  private static final Logger log = LoggerFactory.getLogger(TransactionIdProvider.class);
+
+  public static final long EPOCH_OFFSET_SECONDS = Duration.between(
```
Comment (christydennison) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java]: Maybe keep this in `TransactionId`? Kind of weird to have `TransactionIdProvider` referencing `TransactionId` which references `TransactionIdProvider`

File: soa-commons/src/main/java/com/relateiq/soa/commons/TransactionId.java [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+    this.machineBits = machineBits;
+    this.processBits = processBits;
+    this.counterMask = counterMask;
+  }
+
+  /**
+   * @return The number of seconds since unix epoch
+   */
+  public long getUnixTimestamp() {
+    return timestamp + TransactionIdProvider.EPOCH_OFFSET_SECONDS;
```
Comment (jontg) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionId.java]: I think I agree with @christydennison — lets keep the constant to where it needs to be referenced (either by moving the constant in, or by passing in an offset to `getUnixTimestamp`)

Comment (mjburghoffer) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionId.java]: It is a static field, so it realistically is not scoped to any class at all.  It's just a constant.

File: soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java (hunk 12 above)
Comment (mjburghoffer) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java]: How about the first year of RelateIQ? :D

File: soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java (hunk 13 above)
Comment (mjburghoffer) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java]: good call

File: soa-commons/src/main/java/com/relateiq/soa/commons/TransactionId.java (hunk 15 above)
Comment (mjburghoffer) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionId.java]: good call

File: soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+  private static final Logger log = LoggerFactory.getLogger(TransactionIdProvider.class);
+
+  /*
+  0011111111111111111111111111111111111111111111111111111111111111 WHOLE SHEBANG
+  0011111111111111111111111111111111000000000000000000000000000000 TIMESTAMP
+  0000000000000000000000000000000000111111000000000000000000000000 MACHINE
+  0000000000000000000000000000000000000000111111000000000000000000 PROCESS
+  0000000000000000000000000000000000000000000000111111111111111111 COUNTER
+  */
+  private static final long FULL_TX_MASK = 0b0011111111111111111111111111111111111111111111111111111111111111L;
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/TransactionIdProvider.java]: MAJOR Remove this unused "FULL_TX_MASK" private field. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1068)

**Comment by relateiq-ec2** [comment]
SonarQube analysis reported 8 issues:
* CRITICAL 2 critical
* MAJOR 1 major
* MINOR 5 minor

Watch the comments in this conversation to review them.

File: soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    if (o == this) return true;
+    if (!(o instanceof EndpointInfo)) return false;
+    final EndpointInfo other = (EndpointInfo) o;
+    final Object this$resourcePath = this.resourcePath;
+    final Object other$resourcePath = other.resourcePath;
+    if (this$resourcePath == null ? other$resourcePath != null : !this$resourcePath.equals(other$resourcePath))
+      return false;
+    final Object this$methodPath = this.methodPath;
+    final Object other$methodPath = other.methodPath;
+    if (this$methodPath == null ? other$methodPath != null : !this$methodPath.equals(other$methodPath)) return false;
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java]: MINOR At most one statement is allowed per line, but 2 statements were found on this line. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00122)

File: soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+    final Object this$name = this.name;
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java]: MINOR Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00117)

File: soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+    final Object other$name = other.name;
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java]: MINOR Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00117)

File: soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+    if (this$name == null ? other$name != null : !this$name.equals(other$name)) return false;
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java]: MINOR At most one statement is allowed per line, but 2 statements were found on this line. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00122)

File: soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+    return true;
+  }
+
+  @Override
+  public int hashCode() {
+    final int PRIME = 59;
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java]: MINOR Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00117)

File: soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+    int result = 1;
+    final Object $resourcePath = this.resourcePath;
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java]: MINOR Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00117)

File: soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+    result = result * PRIME + ($resourcePath == null ? 0 : $resourcePath.hashCode());
+    final Object $methodPath = this.methodPath;
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java]: MINOR Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00117)

File: soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+    result = result * PRIME + ($methodPath == null ? 0 : $methodPath.hashCode());
+    final Object $name = this.name;
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java]: MINOR Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00117)

File: soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+
+  public String getResourcePath() {return this.resourcePath;}
+
+  public String getMethodPath() {return this.methodPath;}
+
+  public String getName() {return this.name;}
+
+  @Override
+  public boolean equals(Object o) {
+    if (o == this) return true;
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java]: MINOR At most one statement is allowed per line, but 2 statements were found on this line. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00122)

File: soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+    if (!(o instanceof EndpointInfo)) return false;
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java]: MINOR At most one statement is allowed per line, but 2 statements were found on this line. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00122)

File: soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+    final EndpointInfo other = (EndpointInfo) o;
+    final Object this$resourcePath = this.resourcePath;
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java]: MINOR Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00117)

File: soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+    final Object other$resourcePath = other.resourcePath;
```
Comment (relateiq-ec2) [inline soa-commons/src/main/java/com/relateiq/soa/commons/EndpointInfo.java]: MINOR Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00117)

**Pull request merged**